src/devices/sound/mp3_audio.h

```diff
@@ -0,0 +1,34 @@
+// license:BSD-3-Clause
+// copyright-holders:windyfairy
+/***************************************************************************
+
+    MP3 audio decoder
+
+***************************************************************************/
+
+#ifndef MAME_SOUND_MP3_AUDIO_H
+#define MAME_SOUND_MP3_AUDIO_H
+
+#pragma once
+
+#include <stdint.h>
+
+#include "minimp3/minimp3.h"
+
+class mp3_audio
+{
+public:
+	mp3_audio(const void *base);
+
+	bool decode_buffer(int &pos, int limit, short *output, int &output_samples, int &sample_rate, int &channels);
+
+	void clear();
+
+private:
+	const uint8_t *base;
+
+	mp3dec_t dec;
+	bool m_found_stream;
+};
```

Comment:
Can you forward-declare the `mp3_dec_t`, then make the decoder member a `std::unique_ptr<mp3dec_t>` to avoid needing to include minimp3 headers in the public header here?  If you do that, you may be able to avoid local changes to minimp3 itself as well.

Reply:
It's a typedef unnamed struct so a normal forward declaration wasn't enough and it would error out with an error along the lines of `typedef struct mp3dec_t` is different from `typedef struct mp3dec_t`. I did come up with something that works using inheritance but it may be a little on the hacky side. It does satisfy the requirements of using unique_ptr and does not require any modifications to minimp3.h to work though, and unnecessary info isn't leaked out to the devices.

https://github.com/mamedev/mame/pull/11210/commits/0a60c1fbea5fa3f15b724bcc512e7edd40534e36

Reply:
You could avoid inheritance by forward-declaring the nested structure and making the `mp3dec_t` a member, at the cost of more verbose syntax when using it, like:

In header:
```c++
...
class mp3_audio
{
	...
	struct decoder;
	...
	std::unique_ptr<decoder_t> dec;
	...
};
...
```

In source file:
```c++
...
struct mp3_audio::decoder { mp3dec_t d };
...
void mp3_audio::clear()
{
	mp3dec_init(&dec->d);
	m_found_stream = false;
}
...
```
etc.

But it doesn’t make that much difference.